src/sage/data_structures/list_of_pairs.pyx

```diff
@@ -0,0 +1,118 @@
+r"""
+A data structure to store lists of integer pairs of large size.
+"""
+
+# ****************************************************************************
+#       Copyright (C) 2022 Jonathan Kliem <jonathan.kliem@gmail.com>
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 2 of the License, or
+# (at your option) any later version.
+#                  https://www.gnu.org/licenses/
+# ****************************************************************************
+
+from cysignals.memory cimport check_reallocarray, check_allocarray, sig_free
+from sage.rings.integer cimport smallInteger
+
+# Should be a power of two.
+# Should be neither exposed nor modified.
+cdef size_t length_per_list = 16348
+
+cdef class ListOfPairs:
+    def __dealloc__(self):
+        cdef size_t n_lists = self.length // length_per_list
+        cdef size_t i
+        for i in range(n_lists):
+            sig_free(self._lists[i])
+        sig_free(self._lists)
+
+    cdef inline int enlarge(self) except -1:
+        """
+        Increase size of list by one.
+        """
+        if self.length % length_per_list:
+            self.length += 1
+            return 0
+
+        cdef size_t n_lists = self.length // length_per_list
+        self._lists = <pair_s**> check_reallocarray(self._lists, n_lists + 1, sizeof(pair_s*))
+        self._lists[n_lists] = <pair_s*> check_allocarray(length_per_list, sizeof(pair_s))
+        self.length += 1
+
+    cdef inline pair_s* get(self, size_t index) except NULL:
```

Comment:
Even though it is obvious, it would be good to had a 1-line doc explaining that it returns a pair of the list.

Reply:
Done in 2367bc9685579e0a554a1371f6b919e532ab6845.